File: 17.LetterCombinationsofaPhoneNumber.java
class Solution {
    List<String> mapping = new ArrayList<>(List.of("","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"));
    List<String> result = new ArrayList<>();

    void backtrack(String digits, String letters, int index){

        if(digits.length() == letters.length()){
            result.add(letters);
            return;
        }
        // the "-'0'" -> in digits.charAt(index)-'0' -> converts the character into an integer.
        String currDigit = mapping.get(digits.charAt(index) - '0');
        for (int i = 0; i < currDigit.length(); i++){
            backtrack(digits, letters + currDigit.charAt(i), index + 1);
        }
    }


    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0){
            return result;
        }
        backtrack(digits, "", 0);
        return result;
        
    }
}

File: 17.LetterCombinationsofaPhoneNumber.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        letter_map = {"2":['a','b','c'],"3":["d","e","f"],"4":["g","h","i"],"5":["j","k","l"],"6":["m","n","o"],"7":["p","q","r","s"],"8":["t","u","v"], "9":["w","x","y","z"]}
        res = []
        if digits == "":
            return ""
        def backtrack(digit, curStr):
            if digit == "":
                res.append(curStr)
                return
            
            for c in letter_map[digit[0]]:
                curStr += c
                backtrack(digit[1:], curStr)
                curStr = curStr[:-1]
            
        backtrack(digits, "")
        return res

        
